e2e/packages/sync-test/data/readClientStore.ts

```diff
@@ -2,15 +2,53 @@ import { Page } from "@playwright/test";
 
 // Extract the storeCache type directly from the client
 import { setup } from "../../client-vanilla/src/mud/setup";
+import { deserialize, serialize } from "./utils";
 type StoreCache = Awaited<ReturnType<typeof setup>>["network"]["storeCache"];
 
 /**
  * Read an individual record from the client's data store.
  * This is necessary because `page.evaluate` can only transmit serialisable data,
  * so we can't just return the entire client store (which includes functions to read data)
  */
-export async function readClientStore(page: Page, selector: Parameters<StoreCache["get"]>): Promise<StoreCache> {
-  return page.evaluate((_selector) => {
-    return window["storeCache"].get(..._selector);
+export async function readClientStore(
+  page: Page,
+  [namespace, table, key]: Parameters<StoreCache["get"]>
+): Promise<Record<string, unknown> | undefined> {
+  const selector = [namespace, table, serialize(key)];
+  const serializedValue = await page.evaluate(async (_selector) => {
+    const [_namespace, _table, _key] = _selector;
+    const result = await window["storeCache"].get(_namespace, _table, deserialize(_key));
+    return result ? serialize(result) : undefined;
+
+    /**
+     * Helper to serialize values that are not natively serializable and therefore not transferrable to the page
+     * For now only `bigint` needs serialization.
+     */
+    function serialize(obj: unknown): string {
+      return JSON.stringify(obj, (_, v) => (typeof v === "bigint" ? `bigint(${v.toString()})` : v));
+    }
+
+    /**
+     * Helper to deserialize values that were serialized by `serialize` (because they are not natively serializable).
+     * For now only `bigint` is serialized and need to be deserialized here.
+     */
+    function deserialize(blob: string): Record<string, unknown> {
+      const obj = JSON.parse(blob);
+
+      // Check whether the value matches the mattern `${number}n`
+      // (serialization of bigint in `serialize`)
+      // and turn it back into a bigint
+      const regex = /^bigint\((.*)\)$/; // Regular expression pattern.
+      for (const [key, value] of Object.entries(obj)) {
+        const match = typeof value === "string" && value.match(regex); // Attempt to match the pattern.
+        if (match) {
+          obj[key] = BigInt(match[1]);
+        }
+      }
+
+      return obj;
```

Comment:
might be safer to use the second arg of `JSON.parse` (like we do for `JSON.stringify`) so we account for deep values

```ts
const serializedBigIntPattern = /^bigint\((\d+)n\)$/;

JSON.parse(blob, (key, value) => {
  const matches = typeof value === 'string' && value.match(serializedBigIntPattern);
  if (matches) {
    return BigInt(matches[1]);
  }
  return value;
})
```

Reply:
I realize this code is duplicated by the utils below, so this should prob get moved there.

Reply:
the issue here was that we can't pass functions to the playwright browser, but I just added a workaround that serializes and deserializes those the functions to serialize and deserialize data: https://github.com/latticexyz/mud/pull/1074/commits/1509ddb78589c480e88cdc3c4f4cae2f2429a22a